src/urlencoded.rs

```diff
@@ -30,20 +30,20 @@ impl Params {
 }
 
 pub fn parse(encoded_string : &str) -> Params {
-    Params(form_urlencoded::parse(encoded_string.as_bytes()).into_iter().group())
+    Params(form_urlencoded::parse(encoded_string.as_bytes()).into_owned().group())
 }
 
 pub fn parse_uri(origin: &RequestUri) -> Params {
-    let f = |query: Option<&String>| query.map(|q| parse(&*q));
+    let f = |query: Option<&str>| query.map(|q| parse(&*q));
 
     let result = match *origin {
-        AbsoluteUri(ref url) => f(url.query.as_ref()),
-        AbsolutePath(ref s) => UrlParser::new().parse_path(&*s)
-                                                // FIXME: If this fails to parse,
-                                                // then it really shouldn't have
-                                                // reached here.
-                                               .ok()
-                                               .and_then(|(_, query, _)| f(query.as_ref())),
+        AbsoluteUri(ref url) => f(url.query()),
+        AbsolutePath(ref s) => {
+            let base = "f://".parse().unwrap();
```

Comment:
In this specific case, we can probably just split the string `s` on `?`.

 

> Does an HTTP server want to every edge case the same way? For example, percent-encoding (not decoding) some punctuation/delimiters, normalizing \ to /, normalizing . and .. path segments?

For this, I'm not so sure, perhaps we should raise the issue in `hyper` as it's already using `Url` for a variant in [`RequestUri`](http://hyper.rs/hyper/v0.9.6/hyper/uri/enum.RequestUri.html). cc @seanmonstar